### ISSUE TITLE
[Due for payment 2025-03-24] [$250] Copilot - Copilot with removed access can still edit profile details

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#51678]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.55-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/5140765&group_by=cases:section_id&group_id=229064&group_order=asc
**Email or phone of affected tester (no customers):** applausetester+gm103221@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
Precondition:
1. User A is a main account
2. User B is a copilot with full copilot access
Steps:
1. As User A navigate to Settings > Security
2. Remove Copilot access for User B
3. As User B try to make some changes in profile section (Add display name, status..)

## Expected Result:
Copilot with removed access should not be able to make changes in main account

## Actual Result:
Copilot with removed access can still edit profile details to main account

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/b2dee8bf-13e3-46fe-874c-3692f2815e65

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021854642791837871459</li>
        <li>Upwork Job ID: 1854642791837871459</li>
        <li>Last Price Increase: 2024-11-07</li>
<li>Automatic offers: </li>
<ul>
<li>nkdengineer | Contributor | 104995797</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by lanitochka17** [comment]
@alexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Copilot with removed access can still edit profile details to main account

### What is the root cause of that problem?
The API returns an error as expected but we don't revert the display name or status in failure data


<img width="1512" alt="Screenshot 2024-10-30 at 02 04 11" src="https://github.com/user-attachments/assets/24bb387c-2a70-4b56-9806-f2229ecdd279">


https://github.com/Expensify/App/blob/3edc346319174eba90b494bc1f06c2f28056d4bd/src/libs/actions/User.ts#L1262-L1266
### What changes do you think we should make in order to solve the problem?
We should revert the data in failure data when the API fails when we update the personal details like status, display name, pronouns, and timezone.

https://github.com/Expensify/App/blob/3edc346319174eba90b494bc1f06c2f28056d4bd/src/libs/actions/User.ts#L1262-L1266

### What alternative solutions did you explore? (Optional)
We can display the not found page if the user goes to the update page and the access is removed. We can create a new wrapper page like `DelegateAccessOrNotFoundWrapper` that is the same pattern with `AccessOrNotFoundWrapper`. This wrapper will return not found page if the user is accessing the delegate account and the access is removed.

This can also be used in the feature when we want to display not found page if the page can't be accessed for the limited delegate account.

**Comment by alexpensify** [comment]
Adding to my testing list, I'll review this one soon.

**Comment by alexpensify** [comment]
No update yet

**Comment by alexpensify** [comment]
@s77rt - can you please confirm if this proposal will fix this issue? 

Heads up, I will be offline until Tuesday, November 12, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If it can wait, I'll continue the review process when I return online. Thanks!

**Comment by s77rt** [comment]
🎀 👀 🎀 Needs BE changes

**Comment by s77rt** [comment]
@chiragsalian This may need some BE changes (send pusher events) please check the linked Slack thread above

**Comment by s77rt** [comment]
Discussing on slack. We have a plan, now we need the BE changes and the design

**Comment by alexpensify** [comment]
@s77rt can you share the Slack thread link here? Thanks!

**Comment by s77rt** [comment]
@alexpensify https://expensify.slack.com/archives/C01GTK53T8Q/p1731059122077419

**Comment by s77rt** [comment]
Still discussing

**Comment by s77rt** [comment]
Now, we are looking for proposals...

cc @nkdengineer please update your proposal to reflect the new requirement

**Comment by nkdengineer** [comment]
@s77rt Updated [proposal](https://github.com/Expensify/App/issues/51678#issuecomment-2445119450).

**Comment by s77rt** [comment]
@nkdengineer Thanks for the update!

> When the access is removed, the account data will be changed like this

So the BE already sends us pusher data, no need for another data right? Can you double check the data before and after access is lost? We probably need to use `usePrev` hook so the effect only execute after you have lost access and not when you didn't have access at all

**Comment by s77rt** [comment]
@nkdengineer Looks good but I think we should check for `delegate` in `delegates`, instead of checking the length of delegates as if that account has multiple delegates and you are the only person who got removed the length check will fail.

**Comment by nkdengineer** [comment]
@s77rt Updated [proposal](https://github.com/Expensify/App/issues/51678#issuecomment-2445119450).

**Comment by chiragsalian** [comment]
> Here is the pusher data

Proposal LGTM. Feel free to begin working on the PR @s77rt. 
Regarding this part,
```
If these checks aren't enough to check the delegate is removed or not, BE can send a pusher data for account data with a field in delegatedAccess like isCurrentDelegateAccessRemoved. And we can use this field to detect the access is removed.
{
  delegatedAccess: {
      isCurrentDelegateAccessRemoved: true
  }
}
```
I think before that we can update the push notification to set delegatedAccess.delegate to null right and then this this logic,
```
const isActingAsDelegate = !!account?.delegatedAccess?.delegate ?? false;
```
would because false and be better right? Cause right now with the push notification it looks like delegatedAccess.delegate wont change, correct me if I'm mistaken.

So it sounds like i don't need to make any immediate back end changes because the push notification of,
```
[{"key":"account","onyxMethod":"merge","value":{"delegatedAccess":{"delegates":[],"delegators":[]}}}]
```
should already be sent out. But if otherwise let me know.

**Comment by s77rt** [comment]
@chiragsalian I think we should keep `delegatedAccess.delegate` because technically (from FE perspective) you are still connected to the other account. Another reason is that I think it's better to check if `delegatedAccess.delegate` is not in `delegatedAccess.delegates` rather than checking if `session.email` is not in `delegatedAccess.delegates` because if you have a secondary login the `session.email` would be different than `delegatedAccess.delegate` (or at least that's what I'm expecting). But either way, we don't need BE changes for now.

**Comment by nkdengineer** [comment]
I also tried with `connect` function with email the the current user but the error still the same.

**Comment by s77rt** [comment]
Hmm, okay but can we still show the modal and on user press we log him out? Not ideal but at least the user would be able to use the app again -- (will need to confirm this or find a solution to the session being expired)

**Comment by nkdengineer** [comment]
<img width="601" alt="Screenshot 2024-11-25 at 15 31 18" src="https://github.com/user-attachments/assets/e231ea61-e1a8-4a54-9013-0012bb4dcd17">

@s77rt the current modal content like this. How should we change it?

**Comment by shawnborton** [comment]
I feel like we had this conversation somewhere else too. Basically if UserB is the copilot and suddenly they have their access removed, we should just refresh the page and send UserB back to their own account with an alert modal explaining what happened. Does that sound familiar?

**Comment by dannymcclain** [comment]
Yeah I definitely remember having a conversation somewhere and arriving at that solution. Not sure where, but I think that's the route we should take for sure.

**Comment by s77rt** [comment]
Yeah that's it. We are looking for the copy confirmation or should we ask on slack?

Btw, currently the BE revoke access to the current user and we can't switch back to our account. We will ask the user to logout for now. cc @chiragsalian you may want to look into this if we want to switch account instead of logging the user out

**Comment by shawnborton** [comment]
I would ask on Slack or let's see what @jamesdeanexpensify thinks.

**Comment by jamesdeanexpensify** [comment]
Would this work?

> **Not so fast...**
> You need to be a copilot to access this page. Sorry!

And then maybe link on "copilot" to the copilot [HelpDot page](https://help.expensify.com/articles/new-expensify/settings/Add-or-Act-As-a-Copilot)?

**Comment by shawnborton** [comment]
I think we might actually need copy for when you get booted out of the Copilot account and land suddenly back in your own. Does that sound right @s77rt ?

**Comment by s77rt** [comment]
That was the initial plan but now it seems we can't land the user back on his own account. Instead we want to force user to logout (clicking the button will log you out), and still inform them that their delegate access has been removed.

**Comment by jamesdeanexpensify** [comment]
Something like this then?

> **Copilot access removed**
> You no longer have access to [email/phone] as a [copilot](https://help.expensify.com/articles/new-expensify/settings/Add-or-Act-As-a-Copilot). Sorry!

**Comment by shawnborton** [comment]
That copy works for me 👍

**Comment by chiragsalian** [comment]
Hmm from a product perspective, i think investing time to force the copiloted user to logout is a wrong design choice.

Even if its more complicated we should figure out whats needed so that the copilot user, gets booted out from the copilot account, and back into their original account instead of a logout, and they see the message james provided. Others can weigh in if they feel differently. Unless technically the former is too hard to achieve.

**Comment by shawnborton** [comment]
I agree with that Chirag.

**Comment by s77rt** [comment]
True! The backend invalidates our session and we can't do anything at this point (from frontend perspective). This requires backend changes: If user A is using another account and his access is removed, then sign user A back into his account.

Would you be able to look into this @chiragsalian? Internally I suppose we can just call the signin function but send the onyx data via the pusher

**Comment by chiragsalian** [comment]
> Would you be able to look into this @chiragsalian? 

Sure i can, but full transparency, i may not be able to this week because of more pressing items. I should be able to investigate next week.

**Comment by s77rt** [comment]
@chiragsalian Any updates on this?

**Comment by chiragsalian** [comment]
Sorry, unfortunately not yet. Been busy with higher priority items. I'm hoping to get to this soon.

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Wednesday, December 18, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by alexpensify** [comment]
Weekly Update: Waiting for an internal investigation

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Friday, January 3, 2025, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by alexpensify** [comment]
@chiragsalian - any update if you can review this one soon? Thanks!

**Comment by chiragsalian** [comment]
i don't think there is anything to review here. I basically need to make some backend changes. After that I'm unsure if some external changes would still be necessary. The issue is unfortunately low on my priority list.

**Comment by alexpensify** [comment]
@chiragsalian - I forgot to ask but should we make this a Monthly?

**Comment by chiragsalian** [comment]
i think since this affects users it should remain a weekly. I feel like if we set it to monthly it would just be forgotten and closed.

**Comment by chiragsalian** [comment]
I think i finally get some free cycles this week so i'll have a look at it.

**Comment by chiragsalian** [comment]
Alright i had time to investigate today. I know the current front-end behavior and the backend code. I did research on what oldDot does as well to compare.

It looks like in the backend when userA removes userB as copilot, userB does not get any onyx updates. 
This is why it looks like userB can continue to make changes. Why i say it "looks like" is because userB API requests don't actually succeed (i tried by changing display name). Its just optimistic data getting saved which is why the front end gets updated. The API requests fails with error code `408` and message `The account you are trying to use is deleted.`. 

Alright so how does oldDot handle this. In oldDot, the moment userB takes an action and we see error code 408 we redirect the user to signout and then back to signin with their stashed credentials.

I think we should just do the same here i.e.,
- In the front end when error code 408 is received with that error message then call `signOutAndRedirectToSignIn` and if stashedCredentials and stashedSession exists then use it.

We could get extra fancy and send an onyx update immediately when access is revoked to clear the session key for userB. But then we would need more front end code to handle this scenario. Which imo is not super worth it since the more important flow to handle is the first one mention cause the user may not always get their onyx update. 

So yeah on investigation i think the main part is purely front end and we most likely dont need backend change unless we want to send an onyx update to clear the session key.

What are your thoughts on this @nkdengineer? Do you agree or do you feel like i missed something.

**Comment by chiragsalian** [comment]
> What is stashedCredentials and stashedSession?

Well as the name implies its the stashed credentials and session.
The flow is like this,
1. UserB logs into their own account, at this point in their onyxDB they'll have credentials and session.
  a. Credientials contains - autoGeneratedLogin, autoGeneratedPassword, login, validateCode. These contain information of userB
  b. Session contains - accountID, authToken, creationDate, email, encryptedAuthToken. These contain information of userB.
2. Now when UserB copilots into UserA,
  a. In their onyxDB, their credentials and session are saved into stashedCredentials and stashedSession.
  b. UserB, credentials and session then get updated with the data of UserA.
  c. So at this point for userB - credentials and session identify themselves as userA, stashedCredentials and stashedSession has their information as userB.

Does that help?

So with my suggestion, 
> In the front end when error code 408 is received with that error message then call signOutAndRedirectToSignIn and if stashedCredentials and stashedSession exists then use it.

When 408 is detected if the front end clears the current credentials/session and then moves the data from stashedCredentials/stashedSession back to credentials/session it would bring userB back to their account. Let me know if that makes sense.

**Comment by chiragsalian** [comment]
> While testing this, I see we only use the stashedCredentials and stashedSession for SupportAuthToken. Is this expected?

Well yeah thats the current code on main. We've to either update signOutAndRedirectToSignIn to support this copilot flow, or make a new method to handle moving stashed param to the proper place. My preference would be to update signOutAndRedirectToSignIn, maybe with a bool param like shouldHandleCopilotSignout and it thats true skip the isSupportAuthToken check.

**Comment by alexpensify** [comment]
🚨 Heads up! I'll be offline until Wednesday, March 12, 2025, and won’t be actively monitoring this GitHub during that time. 

If this GitHub requires an urgent update, please reapply the `bug` label or reach out in the `#expensify-open-source` Slack room for help. Otherwise, I'll pick things up when I'm back online. Thanks!  

--- 

Update: The proposal is under review

**Comment by alexpensify** [comment]
@s77rt, can you please review if this warning is accurate or an error? Thanks!

**Comment by s77rt** [comment]
@alexpensify It is a bug but not exactly a regression. Will look into this asap

**Comment by alexpensify** [comment]
@s77rt - any update here? Thanks!

**Comment by s77rt** [comment]
@alexpensify The other bug is being handled in https://github.com/Expensify/App/issues/58465 and we have a PR for that https://github.com/Expensify/App/pull/58538.

**Comment by alexpensify** [comment]
Thank you for the update!

**Comment by alexpensify** [comment]
### Payment SummaryContributor: 
Contributor: @nkdengineer paid $250 via Upwork
Contributor+: @s77rt due $250 via NewDot -- please submit a request

**Comment by s77rt** [comment]
I will request once we are done with https://github.com/Expensify/App/issues/58465 (related issue)

**Comment by JmillsExpensify** [comment]
$250 approved for @s77rt
